libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,721 @@
+    // @User: Standard
+    AP_GROUPINFO("AUTO_SAVE", 11, AP_Quicktune, auto_save, 0),
+
+    // @Param: REDUCE_MAX
+    // @DisplayName: Quicktune maximum gain reduction
+    // @Description: This controls how much quicktune is allowed to lower gains from the original gains. If the vehicle already has a reasonable tune and is not oscillating then you can set this to zero to prevent gain reductions. The default of 20% is reasonable for most vehicles. Using a maximum gain reduction lowers the chance of an angle P oscillation happening if quicktune gets a false positive oscillation at a low gain, which can result in very low rate gains and a dangerous angle P oscillation.
+    // @Units: %
+    // @Range: 0 100
+    // @User: Standard
+    AP_GROUPINFO("REDUCE_MAX", 12, AP_Quicktune, reduce_max, 20),
+
+    // @Param: OPTIONS
+    // @DisplayName: Quicktune options
+    // @Description: Additional options. When the Two Position Switch option is enabled then a high switch position will start the tune, low will disable the tune. you should also set a QUIK_AUTO_SAVE time so that you will be able to save the tune.
+    // @Bitmask: 0:UseTwoPositionSwitch
+    // @User: Standard
+    AP_GROUPINFO("OPTIONS", 13, AP_Quicktune, options, 0),
+
+    AP_GROUPEND
+};
+
+// Call at loop rate
+void AP_Quicktune::update(bool mode_supports_quicktune)
+{
+    if (enable < 1) {
+        if (need_restore) {
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "QuickTune disabled");
+            abort_tune();
+        }
+        return;
+    }
+    const uint32_t now = AP_HAL::millis();
+
+    if (!mode_supports_quicktune) {
+        /*
+          user has switched to a non-quicktune mode. If we have
+          pending parameter changes then revert
+         */
+        if (need_restore) {
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "QuickTune aborted");
+        }
+        abort_tune();
+        return;
+    }
+
+    if (need_restore) {
+        const float att_error = AC_AttitudeControl::get_singleton()->get_att_error_angle_deg();
+        if (att_error > MAX_ATTITUDE_ERROR) {
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "Tuning: attitude error %.1fdeg - ABORTING", att_error);
+            abort_tune();
+            return;
+        }
+    }
+
+    const auto &vehicle = *AP::vehicle();
+
+    if (vehicle.have_pilot_input()) {
+        last_pilot_input = now;
+    }
+
+    SwitchPos sw_pos_tune = SwitchPos::MID;
+    SwitchPos sw_pos_save = SwitchPos::HIGH;
+    if ((options & OPTIONS_TWO_POSITION) != 0) {
+        sw_pos_tune = SwitchPos::HIGH;
+        sw_pos_save = SwitchPos::NONE;
+    }
+
+    if (sw_pos == sw_pos_tune && (!hal.util->get_soft_armed() || !vehicle.get_likely_flying()) && now > last_warning + 5000) {
+        GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "Tuning: Must be flying to tune");
+        last_warning = now;
+        return;
+    }
+    if (sw_pos == SwitchPos::LOW || !hal.util->get_soft_armed() || !vehicle.get_likely_flying()) {
```

Comment:
This seems the be the opposite of the switch positions that just went in for AUTOTUNE in #27754 - I'm not saying which way is better, but having them different is confusing and will cause user frustration.

Reply:
Low switch reverts to the original tune, the behavior is the same.